`odo dev` panics if component name is long (62 characters)

/kind bug

## What versions of software are you using?

**Operating System:**
Fedora 36

**Output of `odo version`:**
```
odo v3.0.0-rc1 (5d4f1b2f0)

Server: https://10.10.10.105:6443
Kubernetes: v1.24.4
```

## How did you run odo exactly?
```
odo init --name document-how-odo-translates-container-component-to-deployment --devfile go --starter go-starter

odo dev
```

## Actual behavior

Validation rules say that name should contain at most 63 characters, which the component name used here meets:

```
❯ yq e '.metadata.name' devfile.yaml | wc -c
62
```

But `odo dev` currently crashes with the following error:

```
❯ odo dev                                                                                                        
  __
 /  \__     Developing using the document-how-odo-translates-container-component-to-deployment Devfile
 \__/  \    Namespace: default
 /  \__/    odo version: v3.0.0-rc1
 \__/

↪ Deploying to the cluster in developer mode
 •  Waiting for Kubernetes resources  ...
panic: runtime error: slice bounds out of range [:31] with length 3

goroutine 1 [running]:
github.com/redhat-developer/odo/pkg/util.NamespaceKubernetesObjectWithTrim({0xc00072f700, 0xc000c14780}, {0x2ea3d5f, 0x1c})
        /home/asoro/work/projects/odo/pkg/util/util.go:122 +0xb0
github.com/redhat-developer/odo/pkg/devfile/adapters/kubernetes/component.(*Adapter).createOrUpdateComponent(0xc000c08ea0, 0x0, {{0xc0006a1ae8, 0x16}, {{{{...}, {...}, {...}, {...}, {...}, {...}, ...}, ...}}}, ...)
        /home/asoro/work/projects/odo/pkg/devfile/adapters/kubernetes/component/adapter.go:451 +0x196a
github.com/redhat-developer/odo/pkg/devfile/adapters/kubernetes/component.Adapter.Push({{0x3306cf8, 0xc0008a9500}, {0x32ee3f0, 0xc0001fef50}, {0x328d640, 0xc00071c7e0}, {0x32eb178, 0xc0009280a0}, {{0xc00072f700, 0x3d}, ...}, ...}, ...)
        /home/asoro/work/projects/odo/pkg/devfile/adapters/kubernetes/component/adapter.go:121 +0x605
github.com/redhat-developer/odo/pkg/dev.(*DevClient).Start(0xc0000c0fa0, {{{0xc000b0a5d0, 0x5}, {0xc0008c5860, 0x16}, {0xc000b0a010, 0xc}, {0xc000a7a900, 0x409, 0x480}, ...}, ...}, ...)
        /home/asoro/work/projects/odo/pkg/dev/dev.go:92 +0x4c5
github.com/redhat-developer/odo/pkg/odo/cli/dev.(*DevOptions).Run(0xc00017d8c0, {0x328c840, 0xc0008d9950})
        /home/asoro/work/projects/odo/pkg/odo/cli/dev/dev.go:201 +0x470
github.com/redhat-developer/odo/pkg/odo/genericclioptions.GenericRun({0x328d1e0, 0xc00017d8c0}, 0xc000aaf400, {0x46ac1c8, 0x0, 0x0})
        /home/asoro/work/projects/odo/pkg/odo/genericclioptions/runnable.go:146 +0xa3e
github.com/redhat-developer/odo/pkg/odo/cli/dev.NewCmdDev.func1(0xc000aaf400, {0x46ac1c8, 0x0, 0x0})
        /home/asoro/work/projects/odo/pkg/odo/cli/dev/dev.go:314 +0x38
github.com/spf13/cobra.(*Command).execute(0xc000aaf400, {0x46ac1c8, 0x0, 0x0})
        /home/asoro/work/projects/odo/vendor/github.com/spf13/cobra/command.go:860 +0x5f8
github.com/spf13/cobra.(*Command).ExecuteC(0xc000930a00)
        /home/asoro/work/projects/odo/vendor/github.com/spf13/cobra/command.go:974 +0x3bc
github.com/spf13/cobra.(*Command).Execute(...)
        /home/asoro/work/projects/odo/vendor/github.com/spf13/cobra/command.go:902
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        /home/asoro/work/projects/odo/vendor/github.com/spf13/cobra/command.go:895
main.main()
        /home/asoro/work/projects/odo/cmd/odo/odo.go:70 +0x425
```

## Expected behavior
It should not panic :-)

## Any logs, error output, etc?

--- a/pkg/util/util.go
+++ b/pkg/util/util.go
@@ -106,27 +106,30 @@ func NamespaceKubernetesObject(componentName string, applicationName string) (st
 // if the resultant name is greater than 63 characters
 // it trims app name then component name
 func NamespaceKubernetesObjectWithTrim(componentName, applicationName string) (string, error) {
+	var (
+		maxLen = 63
+	)
 	value, err := NamespaceKubernetesObject(componentName, applicationName)
 	if err != nil {
 		return "", err
 	}
 
 	// doesn't require trim
-	if len(value) <= 63 {
+	if len(value) <= maxLen {
 		return value, nil
 	}
 
 	// trim app name to max 31 characters
 	appNameLen := len(applicationName)
-	if appNameLen > 31 {
-		appNameLen = 31
+	if appNameLen > maxLen/2 {
+		appNameLen = maxLen / 2
 	}
 	applicationName = applicationName[:appNameLen]
 
 	// trim component name with remaining length
 	minComponentLen := len(componentName)
-	if minComponentLen > 62-appNameLen {
-		minComponentLen = 62 - appNameLen
+	if minComponentLen > maxLen-appNameLen-1 {
+		minComponentLen = maxLen - appNameLen - 1
 	}
 	componentName = componentName[:minComponentLen]
 

--- a/pkg/util/util.go
+++ b/pkg/util/util.go
@@ -104,8 +104,7 @@ func NamespaceKubernetesObject(componentName string, applicationName string) (st
 
 // NamespaceKubernetesObjectWithTrim hyphenates applicationName and componentName
 // if the resultant name is greater than 63 characters
-// it trims each to 31 characters
-// <31-characters>+"-"+<31-characters> = 63 characters
+// it trims app name then component name
 func NamespaceKubernetesObjectWithTrim(componentName, applicationName string) (string, error) {
 	value, err := NamespaceKubernetesObject(componentName, applicationName)
 	if err != nil {

--- a/pkg/util/util.go
+++ b/pkg/util/util.go
@@ -117,9 +117,20 @@ func NamespaceKubernetesObjectWithTrim(componentName, applicationName string) (s
 		return value, nil
 	}
 
-	// trim to 31 characters
-	componentName = componentName[:31]
-	applicationName = applicationName[:31]
+	// trim app name to max 31 characters
+	appNameLen := len(applicationName)
+	if appNameLen > 31 {
+		appNameLen = 31
+	}
+	applicationName = applicationName[:appNameLen]
+
+	// trim component name with remaining length
+	minComponentLen := len(componentName)
+	if minComponentLen > 62-appNameLen {
+		minComponentLen = 62 - appNameLen
+	}
+	componentName = componentName[:minComponentLen]
+
 	value, err = NamespaceKubernetesObject(componentName, applicationName)
 	if err != nil {
 		return "", err

--- a/pkg/util/util_test.go
+++ b/pkg/util/util_test.go
@@ -2571,6 +2571,15 @@ func TestNamespaceKubernetesObjectWithTrim(t *testing.T) {
 			want:    "veryveryveryveryLongComponentGo-veryveryveryveryLongAppNameInGo",
 			wantErr: false,
 		},
+		{
+			name: "case 4: component name is more than 63 characters",
+			args: args{
+				componentName:   "123456789012345678901234567890123456789012345678901234567890ComponentName",
+				applicationName: "app",
+			},
+			want:    "12345678901234567890123456789012345678901234567890123456789-app",
+			wantErr: false,
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {

